[export] fix for duplicate constant lifting

suo opened the pull request:
Summary:
Whenever we access a constant, we emit a `get_attr` node for it.

The `lift_constants_pass` was lifting every `get_attr` node unconditionally, even if the same target was already lifted. This diff fixes that.

I also took the liberty of adding some infra to make it easier to unit test passes. GraphBuilder lets you declaratively construct graphs with the right metadata, it's pretty useful for directly inducing the pattern you want to test against.

Test Plan: added unit test

Differential Revision: D53278161

angelayi reviewed (approved):
very cool graph builder 😎

suo committed:
[export] fix for duplicate constant lifting (#118776)

Summary:

Whenever we access a constant, we emit a `get_attr` node for it.

The `lift_constants_pass` was lifting every `get_attr` node unconditionally, even if the same target was already lifted. This diff fixes that.

I also took the liberty of adding some infra to make it easier to unit test passes. GraphBuilder lets you declaratively construct graphs with the right metadata, it's pretty useful for directly inducing the pattern you want to test against.

Test Plan: added unit test

Reviewed By: khabinov, angelayi

Differential Revision: D53278161
